src/default-request-executor.js

```diff
@@ -33,6 +33,8 @@ class DefaultRequestExecutor extends RequestExecutor {
 
   buildRetryRequest(request, response) {
     const newRequest = deepCopy(request);
+    const delta = this.requestTimeout - (new Date() - request.startTime);
```

Comment:
We also need to account for `delayMs` here, i.e.
- Setup: requestTimeout 10s, first request takes 1s, header returns +5s
- First request 429s, passes checks
- newRequest has a timeout of 9s since it's initialized immediately after the previous response, but will run 5s from now
- should have a timeout of 4s

An easy fix that also accounts for the [timing delays of setTimeout](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/setTimeout#Reasons_for_delays_longer_than_specified) is to modify the calling code:

```javascript
  retryRequest(request, response, delayMs) {
    // Aside from the incorrect timeout, it's better to move this anyway
    // since we're not using them until the retry promise.
    // const newRequest = this.buildRetryRequest(request, response);
    const requestId = this.getOktaRequestId(response);
    return new Promise(resolve => {
      this.emit('backoff', request, response, requestId, delayMs);
      setTimeout(resolve, delayMs);
    }).then(() => {
      const newRequest = this.buildRetryRequest(request, response);
      this.emit('resume', newRequest, requestId);
      return this.fetch(newRequest);
    });
  }
```

Reply:
Good catch, I've updated to account for `delayMs`.  I don't think we should defer the computation of `newRequest.timeout` because as you mention it would definitely mess with our `delta < 1` concern.  Your memory concern is valid, so we could compute the value of `newRequest.timeout` before the `setTimeout` and keep that in the closure, then build `newRequest` after the timeout and attach the timeout value.